### PR TITLE
Allow unlimited number of TCS cabview controls

### PR DESCRIPTION
**Pull request by cesarBLG**
Currently the number of TCS cab view controls is hard-coded to 48. This is not enough when more than one TCS can be active at a given time. Instead of increasing the hard-coded number, this allows an unlimited number of controls in a cleaner way.
https://blueprints.launchpad.net/or/+spec/unlimited-tcs-controls

### CODE REVIEW

File: Source/Orts.Formats.Msts/CabViewFile.cs [hunk 1]
```diff
@@ -356,6 +309,30 @@ public enum DiscreteStates
         CAB_SIGNAL_DISPLAY
     }
 
+    public struct CabViewControlType
+    {
+        public CABViewControlTypes Type;
+        public int Subtype;
```
Comment (Sharpe49) [inline Source/Orts.Formats.Msts/CabViewFile.cs:315]: I think that Id would be a better name.

File: Source/RunActivity/Viewer3D/RollingStock/MSTSLocomotiveViewer.cs [hunk 2]
```diff
@@ -1135,7 +1135,7 @@ public class CabRenderer : RenderPrimitive
         private int _Location;
         private bool _isNightTexture;
         private bool HasCabLightDirectory = false;
-        public Dictionary<int, CabViewControlRenderer> ControlMap;
+        public Dictionary<(CabViewControlType, int), CabViewControlRenderer> ControlMap;
```
Comment (Sharpe49) [inline Source/RunActivity/Viewer3D/RollingStock/MSTSLocomotiveViewer.cs:1138]: Shouldn't CabViewControlType (which is a ref to an object) be a CABViewControlTypes enum value ?

Comment (Sharpe49) [reply Source/RunActivity/Viewer3D/RollingStock/MSTSLocomotiveViewer.cs:1138]: Ok, I was a bit confused by it. Ok.

File: Source/Orts.Formats.Msts/CabViewFile.cs (hunk 1 above)
Comment (cesarBLG) [reply Source/Orts.Formats.Msts/CabViewFile.cs]: Done

**Review by Sharpe49 — APPROVED** [review]

**Comment by cesarBLG** [comment]
Oops! I forgot it

**Comment by cesarBLG** [comment]
Sorry, I did a mistake and pushed to this branch unintentionally

**Comment by Csantucci** [comment]
César,
I have built your branch unlimited-tcs-controls and tested it with the ETR400. When I press Ctrl, (Ctrl + ,), which activates command ControlTCSGeneric1, OR crashes with the following lines in the logfile
Error: System.Collections.Generic.KeyNotFoundException: The given key was not present in the dictionary.
   at System.ThrowHelper.ThrowKeyNotFoundException()
   at System.Collections.Generic.Dictionary`2.get_Item(TKey key)
   at Orts.Viewer3D.RollingStock.MSTSLocomotiveViewer.<InitializeUserInputCommands>b__18_73() in J:\Git_repositories\OR_Cesar\Source\RunActivity\Viewer3D\RollingStock\MSTSLocomotiveViewer.cs:line 193
   at Orts.Viewer3D.RollingStock.MSTSLocomotiveViewer.HandleUserInput(ElapsedTime elapsedTime) in J:\Git_repositories\OR_Cesar\Source\RunActivity\Viewer3D\RollingStock\MSTSLocomotiveViewer.cs:line 277
   at Orts.Viewer3D.RollingStock.MSTSElectricLocomotiveViewer.HandleUserInput(ElapsedTime elapsedTime) in J:\Git_repositories\OR_Cesar\Source\RunActivity\Viewer3D\RollingStock\MSTSElectricLocomotiveViewer.cs:line 75
   at Orts.Viewer3D.Viewer.HandleUserInput(ElapsedTime elapsedTime) in J:\Git_repositories\OR_Cesar\Source\RunActivity\Viewer3D\Viewer.cs:line 901
   at Orts.Viewer3D.Viewer.Update(RenderFrame frame, Single elapsedRealTime) in J:\Git_repositories\OR_Cesar\Source\RunActivity\Viewer3D\Viewer.cs:line 745
   at Orts.Viewer3D.Processes.GameStateViewer3D.Update(RenderFrame frame, Double totalRealSeconds) in J:\Git_repositories\OR_Cesar\Source\RunActivity\Viewer3D\Processes\GameStateViewer3D.cs:line 119
   at Orts.Viewer3D.Processes.UpdaterProcess.Update() in J:\Git_repositories\OR_Cesar\Source\RunActivity\Viewer3D\Processes\UpdaterProcess.cs:line 131
   at Orts.Viewer3D.Processes.UpdaterProcess.DoUpdate() in J:\Git_repositories\OR_Cesar\Source\RunActivity\Viewer3D\Processes\UpdaterProcess.cs:line 108
   at Orts.Viewer3D.Processes.UpdaterProcess.UpdaterThread() in J:\Git_repositories\OR_Cesar\Source\RunActivity\Viewer3D\Processes\UpdaterProcess.cs:line 74
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.ThreadHelper.ThreadStart()

**Comment by cesarBLG** [comment]
Thanks for pointing it, Carlo. It's weird that I didn't spot it, as I've been using these commands a lot for LZB. I've just uploaded a fix.

**Comment by Csantucci** [comment]
It works now, thanks.

**Comment by cesarBLG** [comment]
OK, so now that #713 was merged the problem is with #706. Adding the not-for-unstable tag since the conflicts are preventing the compilation of unstable versions. It's not ideal that we have this obstacles to development with PRs created months ago. Let's see if at least one of the PRs is merged so I can fix the other.

**Review by cjakeman — APPROVED** [review]
This is a very worthwhile refinement.

Approved with thanks,

Chris

**Comment by cesarBLG** [comment]
Thanks, Chris

**Pull request merged**